### DISCUSSION TITLE
[TypeScript] `@octokit/rest` v18.0.5` -> v18.0.6 upgrade breaks usage of Octokit with `@actions/github`

### DISCUSSION THREAD

**Discussion by myyk** [octokit/octokit.js#1872] [Q&A] [answered]
Hi,

I'm not sure if this is a bug or that I'm just new to Typescript and not sure how to use it well.

I found when updating @octokit/rest `18.0.5` -> `18.0.6` that the type of `import {Octokit} from '@octokit/rest' no longer matches 
` that of `import * as github from '@actions/github'` in `github.getOctokit` in `@actions/github` `4.0.0`.

Some code I had like this was failing to compile:

```
    async function foobar(
      octokit: Octokit,
    ): Promise<Foo> {...}

    const octokit = github.getOctokit(inputs.token, {
      previews: ['squirrel-girl']
    }) as Octokit
    foobar(octokit)
```

Is this something that is an actual bug or is there just some simple way to define the types so that it all works?

**Comment by gr2m** [comment]
I'm not aware of any problems here. `@actions/github` is built upon the same libraries as `@octokit/rest` with similar version ranges. I don't see any recent changes in `@actions/github` either.

I tried the folloiwng and didn't get any error

```
import { Octokit } from "@octokit/rest";
import * as github from "@actions/github";

async function foobar(octokit: Octokit): Promise<string> {
  return "test";
}

const octokit = github.getOctokit("123", {
  previews: ["squirrel-girl"],
}) as Octokit;
foobar(octokit);
```

Using the latest `@octokit/rest` and `@actions/github`

```
➜ npm ls @octokit/rest
└── @octokit/rest@18.0.6 
➜ npm ls @actions/github
└── @actions/github@4.0.0
```

**Comment by myyk** [comment]
Hi @gr2m , thanks for your reply and trying to repro the issue.

I double checked and created this branch with only the update to that package and I'm still seeing the same issue: https://github.com/myyk/git-democracy/compare/update-deps-2 can be seen with `npm run build`.

```
src/main.ts:36:21 - error TS2352: Conversion of type '{ [x: string]: any; } & { [x: string]: any; } & Octokit &
 RestEndpointMethods & { paginate: PaginateInterface; }' to type '{ [x: string]: any; } & { [x: string]: any; } &
 Octokit & void & { paginate: PaginateInterface; } & RestEndpointMethods' may be a mistake because
 neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
  Type '{ [x: string]: any; } & { [x: string]: any; } & Octokit & RestEndpointMethods & { paginate: PaginateInterface; }'
 is not comparable to type 'void'.

36     const octokit = github.getOctokit(inputs.token, {
                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
37       previews: ['squirrel-girl']
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
38     }) as Octokit
   ~~~~~~~~~~~~~~~~~
```

I checked that I'm using the same latest dependencies as you:
Using the latest @octokit/rest and @actions/github
```
➜ npm ls @octokit/rest
└── @octokit/rest@18.0.6 
➜ npm ls @actions/github
└── @actions/github@4.0.0
```

Everything seems to compile again if I use `as unknown as Octokit` but it seems I shouldn't need to ignore all type safety.

-------------------------
I just tried to reduce my project even further to just the code you shared: https://github.com/myyk/git-democracy/pull/new/min-project
Using the same code that was working for you, it's not compiling for me. I'm not sure what setting is causing the issue. Most of my project is the Typescript Github Action template.

**Reply by gr2m** [reply]
Okay thanks, I'll have a look. I'm curious where the `void` is suddenly coming from that causes the type conflict ... 🤔

**Reply by gr2m** [reply]
Just to keep note of my debugging:

[![rubber duck](https://user-images.githubusercontent.com/39992/93249657-80dd9c80-f746-11ea-8460-57e5cd3ab5fc.png)](https://www.mindovermachines.com/when-you-get-stuck-talk-to-the-duck/)
When I install `@octokit/rest` v18.0.5 in the `git-democracy` project, the error goes away. But the `void` does not seem to be the problem, the `octokit` instance still has it

<img width="519" alt="image" src="https://user-images.githubusercontent.com/39992/93249185-c3eb4000-f745-11ea-851e-9fbf3046e8d8.png">

So the problem must lie somewhere else. The PR that triggered the 18.0.5 release is https://github.com/octokit/rest.js/pull/1871. It bumped `@octokit/plugin-rest-endpoint-methods` from v4.1.4 to v4.2.0. The v4.2.0 release was triggered by this PR: https://github.com/octokit/plugin-rest-endpoint-methods.js/pull/217. This PR introduced new `octokit.codeScanning.*` methods, and also deprecated the `alert_id` parameter, which is now `alert_number`: https://github.com/octokit/plugin-rest-endpoint-methods.js/pull/217/files?file-filters%5B%5D=.ts#diff-87fa08ed5d6ea8b6887de01903f0c2f8

This PR also bumped `@octokit/types` to v5.5.0: https://github.com/octokit/plugin-rest-endpoint-methods.js/pull/217/files?file-filters%5B%5D=.json#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R27. The pull request that triggerd `@octokit/types` v5.5.0 is https://github.com/octokit/types.ts/pull/165. This pull request was special, because it introduced the deprecation of endpoint paths in case URL parameter change, like it did with the `alert_id` parameter, see https://github.com/octokit/types.ts/pull/165/files#diff-cc9ecb42a0f8d0e8091cd7338ac6d8a0R2260-R2276. It was a more complex `@octokit/types` release than others because of it, so it is possible that a bug slipped in that causes the Type Problems 

What I found is that the new version of `@octokit/types @myyk found.

When setting up https://github.com/myyk/git-democracy/, checking out `update-deps-2`, running npm install and then `npm ls @octokit/types` I see the following

```
npm ls @octokit/types
typescript-action@0.0.0 /Users/gregor/Projects/repro/octkit/rest.js/1872-git-democracy
├─┬ @actions/github@4.0.0
│ ├─┬ @octokit/core@3.1.2
│ │ ├─┬ @octokit/auth-token@2.4.2
│ │ │ └── @octokit/types@5.4.1  deduped
│ │ ├─┬ @octokit/graphql@4.5.4
│ │ │ └── @octokit/types@5.4.1  deduped
│ │ ├─┬ @octokit/request@5.4.7
│ │ │ ├─┬ @octokit/endpoint@6.0.5
│ │ │ │ └── @octokit/types@5.4.1  deduped
│ │ │ ├─┬ @octokit/request-error@2.0.2
│ │ │ │ └── @octokit/types@5.4.1  deduped
│ │ │ └── @octokit/types@5.4.1  deduped
│ │ └── @octokit/types@5.4.1 
│ ├─┬ @octokit/plugin-paginate-rest@2.3.1
│ │ └── @octokit/types@5.4.1  deduped
│ └─┬ @octokit/plugin-rest-endpoint-methods@4.1.3
│   └── @octokit/types@5.4.1  deduped
└─┬ @octokit/rest@18.0.6
  └─┬ @octokit/plugin-rest-endpoint-methods@4.2.0
    └── @octokit/types@5.5.0
```

There are two different versions of `@octokit/types@5.5.0`, which might be the cause of the problem, that's what I'll investigate next

**Reply by gr2m** [reply]
looks like if found the problem: https://github.com/octokit/rest.js/discussions/1872#discussioncomment-79407

thank you [Rubber Duckie](https://youtu.be/Mh85R-S-dh8?t=27)

**Comment by gr2m** [comment, marked as answer]
The problem seems to be a conflict between `@octokit/types@5.5.0` and `@octokit/types@5.4.1`, both versions are present in the dependency tree, see https://github.com/octokit/rest.js/discussions/#discussioncomment-79071

What resolved the problem for me is to update the lock file

```
rm -rf package-lock.json node_modules
npm install
```

After that, the build succeeds.

I wonder, why do you have both `@actions/github` and `@octokit/rest` in your dependencies? It should not be necessary. If you didn't, the problem would not have occurred.

**Reply by myyk** [reply]
Thank you @gr2m! I don't know why I have both `@actions/github` and `@octokit/rest` my dependencies. This is literally my first typescript project (and javascript project) so I thought that I needed `@octokit/rest` and it wasn't obvious to me from any examples I found on usage of the API that I could similarly use: `import {GitHub} from '@actions/github/lib/utils'` in place of `import {Octokit} from '@octokit/rest'`. It wasn't super obvious to me.

I just got everything working. There was a bit more needed to remove `@octokit/rest` like you said, so this is what I did if anyone else is interested:

```
nmp uninstall @octokit/rest
rm -rf package-lock.json node_modules
npm install
```
Then find replace `import {Octokit} from '@octokit/rest'` to `import {GitHub as Octokit} from '@actions/github/lib/utils'`.
